src/p11_child/p11_child_openssl.c

```diff
@@ -42,8 +43,325 @@ struct p11_ctx {
+        if (rv == 0) {
+            DEBUG(SSSDBG_OP_FAILURE, "Timeout on connect\n");
+            return NULL;
+        }
+    }
+
+    ctx = OCSP_sendreq_new(cbio, path, NULL, -1);
+    if (ctx == NULL) {
+        return NULL;
+    }
+
+    if (OCSP_REQ_CTX_add1_header(ctx, "Host", host) == 0) {
+        goto err;
+    }
+
+    if (!OCSP_REQ_CTX_set1_req(ctx, req)) {
+        goto err;
+    }
+
+    for (;;) {
+        rv = OCSP_sendreq_nbio(&rsp, ctx);
+        if (rv != -1)
+            break;
+        if (req_timeout == -1)
+            continue;
+        FD_ZERO(&confds);
+        FD_SET(fd, &confds);
+        tv.tv_usec = 0;
+        tv.tv_sec = req_timeout;
+        if (BIO_should_read(cbio)) {
+            rv = select(fd + 1, (void *)&confds, NULL, NULL, &tv);
+        } else if (BIO_should_write(cbio)) {
+            rv = select(fd + 1, NULL, (void *)&confds, NULL, &tv);
+        } else {
+            DEBUG(SSSDBG_OP_FAILURE, "Unexpected retry condition\n");
+            goto err;
+        }
+        if (rv == 0) {
+            DEBUG(SSSDBG_OP_FAILURE, "Timeout on request\n");
+            break;
+        }
+        if (rv == -1) {
+            DEBUG(SSSDBG_OP_FAILURE, "Select error\n");
+            break;
+        }
+
+    }
+ err:
+    OCSP_REQ_CTX_free(ctx);
+
+    return rsp;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+#define TLS_client_method SSLv23_client_method
+#define OCSP_REQUEST_add0_id OCSP_request_add0_id
+#endif
+
+OCSP_RESPONSE *process_responder(OCSP_REQUEST *req,
+                                 const char *host, const char *path,
+                                 const char *port, int use_ssl,
+                                 int req_timeout)
+{
+    BIO *cbio = NULL;
+    SSL_CTX *ctx = NULL;
+    OCSP_RESPONSE *resp = NULL;
+
+    cbio = BIO_new_connect(host);
+    if (cbio == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Error creating connect BIO\n");
+        goto end;
+    }
+    if (port != NULL)
+        BIO_set_conn_port(cbio, port);
+    if (use_ssl == 1) {
+        BIO *sbio;
+        ctx = SSL_CTX_new(TLS_client_method());
+        if (ctx == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Error creating SSL context.\n");
+            goto end;
+        }
+        SSL_CTX_set_mode(ctx, SSL_MODE_AUTO_RETRY);
+        sbio = BIO_new_ssl(ctx, 1);
+        cbio = BIO_push(sbio, cbio);
+    }
+
+    resp = query_responder(cbio, host, path, req, req_timeout);
+    if (resp == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Error querying OCSP responder\n");
+    }
+
+ end:
+    BIO_free_all(cbio);
+    SSL_CTX_free(ctx);
+    return resp;
+}
+
+static errno_t do_ocsp(struct p11_ctx *p11_ctx, X509 *cert)
+{
+    OCSP_REQUEST *ocsp_req = NULL;
+    OCSP_RESPONSE *ocsp_resp = NULL;
+    OCSP_BASICRESP *ocsp_basic = NULL;
+    OCSP_CERTID *cid = NULL;
+    STACK_OF(OPENSSL_STRING) *ocsp_urls = NULL;
+    char *url_str;
+    X509 *issuer = NULL;
+    int req_timeout = -1;
```

Comment:
I don't see where is req_timeout set, is all the handling for req_timeout != -1 for future extension?

Reply:
Yes, currently the PAM responder handles timeouts by just killing p11_child. But my plan is to add an option to sssctl to validate certificates and here I'd prefer to use a timeout instead of hard killing p11_child.